### ISSUE TITLE
use hashmap instead of BTreeMap for faster lookup

### ISSUE THREAD

**Issue by malik672** [paradigmxyz/reth#12412]
### Describe the feature

I'm not entirely sure but in the but in the pending file under transaction_pool crate we can already order using the the sequential structure of the submission id hence we do not need to use it properties again and we can just switch to a hashmap instead and enable faster lookup constant time than logarithmic time

### Additional context
``` #[derive(Debug, Clone)]
pub struct PendingPool<T: TransactionOrdering> {
    /// How to order transactions.
    ordering: T,
    /// Keeps track of transactions inserted in the pool.
    ///
    /// This way we can determine when transactions were submitted to the pool.
    submission_id: u64,
    /// _All_ Transactions that are currently inside the pool grouped by their identifier.
    by_id: BTreeMap<TransactionId, PendingTransaction<T>>,
    /// _All_ transactions sorted by priority
    all: BTreeSet<PendingTransaction<T>>,
    /// The highest nonce transactions for each sender - like the `independent` set, but the
    /// highest instead of lowest nonce.
    ///
    /// Sorted by their scoring value.
    highest_nonces: BTreeSet<PendingTransaction<T>>,
    /// Independent transactions that can be included directly and don't require other
    /// transactions.
    ///
    /// Sorted by their scoring value.
    independent_transactions: BTreeSet<PendingTransaction<T>>,
    /// Keeps track of the size of this pool.
    ///
    /// See also [`PoolTransaction::size`](crate::traits::PoolTransaction::size).
    size_of: SizeTracker,
    /// Used to broadcast new transactions that have been added to the `PendingPool` to existing
    /// `static_files` of this pool.
    new_transaction_notifier: broadcast::Sender<PendingTransaction<T>>,
}
```

_No response_

**Comment by mattsse** [comment]
hmm, looks like we can do this for the by_id map at least because this function is not a hot path and we can simply iterate over the entire map for this

https://github.com/paradigmxyz/reth/blob/0da914eaebe5fc296098e3b2f406c7250ef2abd4/crates/transaction-pool/src/pool/pending.rs#L519-L524

**Comment by malik672** [comment]
Agreed, since by_id is also used for direct lookups which are more frequent, we can optimize those to O(1) with HashMap while accepting the O(n) scan for get_txs_by_sender

**Comment by malik672** [comment]
I should benchmark this tbh

**Comment by malik672** [comment]
@mattsse i want to remove ```BTreeSet<PendingTransaction<T>>```, I feel it's redundant and not needed since it mirrors ```by_id```, if we want all, we can simply create a helper function over by_id and get all

the best part, we save error, and reduce unnecessary writes

**Comment by hai-rise** [comment]
We also found this in our block-building hot path, especially around cloning `by_id` in `PendingPool::best`:

![Image](https://github.com/user-attachments/assets/a7aa69d7-fef5-4d24-b981-4c11f1a53593)

And dropping `BestTransactions`:

![Image](https://github.com/user-attachments/assets/be68da8a-5741-45ff-a915-023c9f70b4c5)

24-27% of block-building time is spent on cloning and dropping `BestTransactions`... We're writing our own mempool but for upstream, replacing the `BTreeMap` here may still be worth it.

A quick benchmark for cloning and dropping 1 million of `TransactionId -> PendingTransaction` entries shows:
- `Vec`: 16ms
- `HashMap`: 49ms
- `BTreeMap`: 68ms

Of course, we still need to balance it with query complexity and ask if there would be millions of transactions in the mempool. For us, we hope there will be!

**Comment by Houseofkholin** [comment]
@mattsse I’ve been locked out of my GitHub acc, so can’t really work on this right now till the next 3 days
